jfix-stdlib-concurrency/src/main/java/ru/fix/stdlib/concurrency/futures/PendingFutureLimiter.java

```diff
@@ -0,0 +1,288 @@
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+
+/**
+ * maxPendingCount - max amount of pending futures could be registered in limiter instance
+ *
+ * @author Kamil Asfandiyarov
+ */
+public class PendingFutureLimiter {
+    private final Logger log = LoggerFactory.getLogger(PendingFutureLimiter.class);
+
+    private final AtomicLong counter = new AtomicLong();
+    private final ConcurrentHashMap<CompletableFuture, Long> pendingCompletableFutures = new ConcurrentHashMap<>();
+    private volatile int maxPendingCount;
+    private volatile float thresholdFactor = 0.3f;
+    private volatile long maxFutureExecuteTime;
+    private volatile long waitTimeToCheckSizeQueue = TimeUnit.MINUTES.toMillis(1);
+
+    private ThresholdListener thresholdListener;
+
+    /**
+     * @param maxPendingCount         max amount of {@link CompletableFuture} started in parallel
+     * @param maxFutureExecuteTimeout maximum time (milliseconds) on {@link CompletableFuture} execution,
+     *                                when this timeout is reached, {@link CompletableFuture} queue will be cleaned,
+     *                                and on the {@link CompletableFuture} itself
+     *                                {@link CompletableFuture#completeExceptionally(Throwable)} will be invoked
+     *
+     */
+    public PendingFutureLimiter(int maxPendingCount, long maxFutureExecuteTimeout) {
+        this.maxPendingCount = maxPendingCount;
+        this.maxFutureExecuteTime = maxFutureExecuteTimeout;
+    }
+
+    /**
+     * @param maxPendingCount         max amount of {@link CompletableFuture} started in parallel
+     * @param maxFutureExecuteTimeout maximum time (milliseconds) on {@link CompletableFuture} execution,
+     *                                when this timeout is reached, {@link CompletableFuture} queue will be cleaned,
+     *                                and on the {@link CompletableFuture} itself
+     *                                {@link CompletableFuture#completeExceptionally(Throwable)} will be invoked
+     */
+    public PendingFutureLimiter(DynamicProperty<Integer> maxPendingCount, DynamicProperty<Long> maxFutureExecuteTimeout) {
+        this.maxPendingCount = maxPendingCount.get();
+        this.maxFutureExecuteTime = maxFutureExecuteTimeout.get();
+
+        maxPendingCount.addListener((oldValue, newValue) -> setMaxPendingCount(newValue));
+        maxFutureExecuteTimeout.addListener((oldValue, newValue) -> setMaxFutureExecuteTime(newValue));
+    }
+
+    private static int calculateThreshold(int maxPendingCount, float thresholdFactor) {
+        return (int) (maxPendingCount * (1 - thresholdFactor));
+    }
+
+    protected int getThreshold() {
+        return calculateThreshold(maxPendingCount, thresholdFactor);
+    }
+
+    protected int getMaxPendingCount() {
+        return maxPendingCount;
+    }
+
+    /**
+     * Изменяет максимальное число pending futures
```

Comment:
better to translate documentation to english

Reply:
Translated